### ISSUE TITLE
question about backbone ‘Swin’

### ISSUE THREAD

**Issue by Kaacoinnn** [amaralibey/MixVPR#21]
Hi, there. Thanks for your amazing work！And I have some questions about the hyper-parameters in training with 'swin' backbone.
Would you please to show that the hyper-parameter (learning rate, etc. ) when using Swin transformer? 

I replaced the backbone structure in the network with a VIT, which uses the same learning rate as training resnet. But the recall is very bad. I haven't frozen some weights... Do you have any suggestions on this?

**Comment by amaralibey** [comment]
Hello @Kaacoinnn,

We provided Swin as the backbone in this repo and implemented a strategy of freezing most of the first blocks during training. 
To use it, you need to change the following arguments:

```
image_size=(256, 256), # swin takes input images of size 256x256
...
backbone_arch='swin', # call swin backbone
...
agg_config={
                # MixVPR takes in feature maps of size (batch_size, in_channels, in_h, in_w)
                # fill the following according the backbone used
                'in_channels' : 1024,
                'in_h' : 8, # swin outputs 64 tokens which we resize to 8x8 to meet MixVPR requierment
                'in_w' : 8,
                
                # MixVPR configuration
                'mix_depth' : 4,
                'mlp_ratio' : 1,
                'out_channels' : 1024,
                'out_rows' : 4}, # the output dim will be (out_rows * out_channels)
```

Additionally, when using smaller batch sizes, it is important to adjust the learning rate proportionally. For instance, if we set the learning rate to 0.05 when the batch size is 120, halving the batch size to 60 would require reducing the learning rate by half. 
A suitable value for the adjusted learning rate can be calculated using the formula sgd_lr = 0.05 * batch_size / 120. adamw_lr = 0.0002 * batch_size / 120. 

You also have the option to use ViT as the backbone. However, it is important to freeze most of the first blocks and adjust the learning rate based on your specific batch size.

Note that using transformers yiel in very high recall values. However, we didn't include that in our paper because of the latency and the collossal number of parameters associated with transformers.

**Comment by Kaacoinnn** [comment]
very thanks for replying. I will conduct the experiment according to your suggestion~ @amaralibey

**Comment by Kaacoinnn** [comment]
Hi~ @amaralibey 
Hello, first I validated the performance of using Swin Transformer as the backbone, with a recall@1 of 92.3 on the Pittsburgh30k dataset and a recall@1 of 76 on MSLS_val. 

Then I used VIT as the backbone network, as you suggested, with a learning rate of lr=0.0125 (`batch size is 30; 0.05×30/120`) and froze the first six layers of VIT. 

However, after conducting experiments, I found that using VIT as the backbone network only resulted in a recall@1 of 50 on MSLS_val. The recall of 50 was achieved in epoch 7, and the accuracy did not improve in the following 72 epochs of training, as shown in the figure. 

So I am not sure if the issue is with the learning rate setting or the loss function not providing effective constraints, leading to this result. Do you have any suggestions regarding this issue?

![99A AA(TUPTFYVS 9IZ{SX8](https://github.com/amaralibey/MixVPR/assets/76690432/4bfa7f69-aed8-496a-ae83-d066df438eb8)
![9GPLH95UJ3I1WO2YBW%`C%1](https://github.com/amaralibey/MixVPR/assets/76690432/bd93b559-9e8f-4f09-b07b-15e1e530b1a8)

**Comment by amaralibey** [comment]
Hello @Kaacoinnn,

I just tried training with Swin backbone for a single epoch, with batch size of 30 places. I got `92.4% R@1` on `pitts30k_val` and `83.5%` on `msls_val`. I used `adamw` optimizer (with lr == 30*0.0002/120, which is `5e-5`, and `1500`  warmup steps). I guess performance will improve with more epochs.

As for ViT, I think the reason you're getting weak performance is that ViT has been trained with a different mean and std preprocessing.  To address this issue, you can modify the mean_std parameter in the dataloader code at the following line [ LINK](https://github.com/amaralibey/MixVPR/blob/31de0c35eef2c82589481293a200c9ff61741eb5/dataloaders/GSVCitiesDataloader.py#L54), to `mean_std=VIT_MEAN_STD`. 

I hope this helps,
Best regards

**Comment by Kaacoinnn** [comment]
Thank you for your advice！I will conduct another experiment to see its effectiveness.^ ^
